File: backend/routes/signing.py
import uuid
import time
import asyncio
from fastapi import APIRouter, HTTPException
from store.keystore import keystore, HAS_OQS
from services.crypto_service import sign_component
from models.schemas import Component

router = APIRouter()


@router.post("/keygen")
async def generate_keypair():
    result = keystore.generate_keypair()
    return {
        **result,
        "message": "ML-DSA-65 keypair generated" + (" via liboqs (REAL)" if HAS_OQS else " (high-fidelity simulation)"),
        "ml_dsa_details": {
            "algorithm": "ML-DSA-65",
            "standard": "NIST FIPS 204",
            "security_level": "NIST Level 3",
            "hard_problem": "Module-LWE + Module-SIS",
            "public_key_size": 1952,
            "private_key_size": 4032,
            "signature_size": 3293,
            "lattice_dimension": 1024,
        },
        "ed25519_details": {
            "algorithm": "Ed25519",
            "public_key_size": 32,
            "signature_size": 64,
            "quantum_safe": False,
        },
    }


@router.post("/sign-all")
async def sign_all_components(data: dict):
    if not keystore.has_keys():
        raise HTTPException(
            status_code=400,
            detail="Generate keypair first via /latticeguard-api/sign/keygen",
        )

    signed_components = []
    for item in data.get("components", []):
        comp = Component(**item["component"])
        cves_raw = item.get("cves", [])

        from models.schemas import CVE
        cves = []
        for c in cves_raw:
            try:
                cves.append(CVE(**c))
            except Exception:
                pass

        signed = sign_component(comp, cves)
        signed_components.append(signed.model_dump())
        await asyncio.sleep(0.05)

    serial = f"urn:uuid:{uuid.uuid4()}"
    generated_at = str(time.time())

    # Record SBOM signing in audit log
    from store.audit_store import audit_store
    audit_store.add_event(
        "SBOM_SIGNED",
        f"SBOM signed with ML-DSA-65 + Ed25519 — {len(signed_components)} component(s), "
        f"serial {serial}",
        {
            "serial_number": serial,
            "component_count": len(signed_components),
            "algorithm": "Hybrid(Ed25519 + ML-DSA-65)",
            "fips_standard": "FIPS-204",
            "security_level": "NIST-Level-3",
            "ml_dsa_sig_size": 3293,
            "ed25519_sig_size": 64,
        },
    )

    return {
        "bom_format": "CycloneDX",
        "spec_version": "1.5",
        "serial_number": serial,
        "generated_at": generated_at,
        "tool": "LatticeGuard-v2.0",
        "components": signed_components,
        "total_components": len(signed_components),
        "quantum_safe": True,
        "algorithm": "Hybrid(Ed25519 + ML-DSA-65)",
        "fips_standard": "FIPS-204",
    }

File: artifacts/latticeguard-sbom/src/api/signing.ts
import { api } from "./client";
import type { ResolvedPackage } from "./packages";

export interface KeygenResult {
  ml_dsa_public_key: string;
  ml_dsa_public_key_size: number;
  ed25519_public_key: string;
  algorithm: string;
  security_level: string;
  fips_standard: string;
  generated_at: number;
  using_real_oqs: boolean;
  message: string;
  ml_dsa_details: {
    algorithm: string;
    standard: string;
    security_level: string;
    hard_problem: string;
    public_key_size: number;
    private_key_size: number;
    signature_size: number;
    lattice_dimension: number;
  };
}

export interface SignedComponent {
  component: {
    name: string;
    version: string;
    ecosystem: string;
    purl: string;
    description: string;
    author: string;
    license: string;
    homepage: string;
    sha256: string;
    size_bytes: number;
    upload_date: string;
    dependencies: string[];
    depth: number;
  };
  sha256_signed: string;
  ed25519_signature: string;
  ml_dsa_signature: string;
  public_key_ed25519: string;
  public_key_ml_dsa: string;
  algorithm: string;
  fips_standard: string;
  security_level: string;
  signed_at: string;
  signature_size_bytes: number;
  cves: Array<{
    id: string;
    severity: string;
    summary: string;
    fixed_in?: string;
    published: string;
  }>;
}

export interface SBOM {
  bom_format: string;
  spec_version: string;
  serial_number: string;
  generated_at: string;
  tool: string;
  components: SignedComponent[];
  total_components: number;
  quantum_safe: boolean;
  algorithm: string;
  fips_standard: string;
}

export function generateKeypair(): Promise<KeygenResult> {
  return api.post("/sign/keygen", {});
}

export function signAllComponents(
  components: ResolvedPackage[]
): Promise<SBOM> {
  return api.post("/sign/sign-all", { components });
}
